File: ripperdoc/dock/jetbrains-ide/toolbox/read_logs.py
#!/usr/bin/env python3
"""Read the last N lines of the running JetBrains IDE's log.

Usage:  read_logs.py [n]   (default 50)
"""

import argparse
import sys

from core import JetBrainsError, resolve_log_dir


def read_logs(n: int = 50) -> list[str]:
    """Last `n` lines of the active IDE's idea.log.

    `resolve_log_dir()` is the single guard for a live IDE: it raises
    JetBrainsError if none is found.
    """
    log = resolve_log_dir() / "idea.log"
    return log.read_text(errors="replace").splitlines()[-n:]


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        prog="read_logs.py", description="Read the last N lines of the running JetBrains IDE's log."
    )
    parser.add_argument("n", type=int, nargs="?", default=50, help="number of trailing log lines to print (default 50)")
    ns = parser.parse_args(argv)
    try:
        lines = read_logs(ns.n)
    except (JetBrainsError, OSError) as exc:
        print(f"read_logs: {exc}", file=sys.stderr)
        return 1
    print("\n".join(lines))
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: ripperdoc/dock/idea/toolbox/read-logs.ts
#!/usr/bin/env bun
import * as fs from "node:fs/promises"
import * as path from "node:path"
import { defineCommand, execute } from "cmdore"
import { assertIdea } from "./assert-idea.ts"
import { resolveLogDir } from "./core"
import { integer } from "./core/coercers.ts"

/**
 * Split `text` into lines on \r\n / \n / \r, dropping the trailing empty
 * segment after a final terminator (so "a\nb\n" -> ["a", "b"]).
 */
const splitLines = (text: string): string[] => {
    if (text === "") {
        return []
    }
    const lines = text.split(/\r\n|\r|\n/)
    if (lines.length > 0 && lines[lines.length - 1] === "") {
        lines.pop()
    }
    return lines
}

/**
 * Last `n` lines of the active IDE's idea.log. Read as latin1 so every byte
 * decodes without throwing.
 *
 * @param n - line count: n>0 -> last n; n==0 -> all; n<0 -> drop the first |n|.
 * @returns the selected log lines, terminators stripped.
 *
 * @example
 * const tail = await readLogs() // last 50 lines
 * const all = await readLogs(0) // the whole log
 */
export const readLogs = async (n = 50): Promise<string[]> => {
    const log = path.join(await resolveLogDir(), "idea.log")
    const lines = splitLines(await fs.readFile(log, { encoding: "latin1" }))
    return lines.slice(n > 0 ? Math.max(0, lines.length - n) : -n)
}

const command = defineCommand({
    name: "read-logs",
    description: "Read the last N lines of the running JetBrains IDE's log.",
    arguments: [{ name: "n", description: "number of lines to read (default 50)", coerce: integer }],
    run: async ({ n }) => {
        assertIdea()
        const lines = await readLogs(n ?? 50)
        console.log(lines.join("\n"))
    }
})

if (import.meta.main) {
    process.exit(await execute(command, { metadata: command }))
}
